pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -519,322 +519,331 @@ func ValidateVirtualMachineInstanceSpec(field *k8sfield.Path, spec *v1.VirtualMa
+					forwardPort.Protocol = "TCP"
+				}
 
-					if forwardPort.Port < 0 || forwardPort.Port > 65536 {
+				if forwardPort.Name != "" {
+					if _, ok := portForwardMap[forwardPort.Name]; ok {
 						causes = append(causes, metav1.StatusCause{
-							Type:    metav1.CauseTypeFieldValueInvalid,
-							Message: "Port field must be in range 0 < x < 65536.",
-							Field:   field.Child("domain", "devices", "interfaces").Index(idx).Child("ports").Index(portIdx).String(),
+							Type:    metav1.CauseTypeFieldValueDuplicate,
+							Message: fmt.Sprintf("Duplicate name of the port: %s", forwardPort.Name),
+							Field:   field.Child("domain", "devices", "interfaces").Index(idx).Child("ports").Index(portIdx).Child("name").String(),
 						})
 					}
 
-					if forwardPort.Protocol != "" {
-						if forwardPort.Protocol != "TCP" && forwardPort.Protocol != "UDP" {
-							causes = append(causes, metav1.StatusCause{
-								Type:    metav1.CauseTypeFieldValueInvalid,
-								Message: "Unknown protocol, only TCP or UDP allowed",
-								Field:   field.Child("domain", "devices", "interfaces").Index(idx).Child("ports").Index(portIdx).Child("protocol").String(),
-							})
-						}
-					} else {
-						forwardPort.Protocol = "TCP"
+					if msgs := validation.IsValidPortName(forwardPort.Name); len(msgs) != 0 {
+						causes = append(causes, metav1.StatusCause{
+							Type:    metav1.CauseTypeFieldValueInvalid,
+							Message: fmt.Sprintf("Invalid name of the port: %s", forwardPort.Name),
+							Field:   field.Child("domain", "devices", "interfaces").Index(idx).Child("ports").Index(portIdx).Child("name").String(),
+						})
 					}
 
-					if forwardPort.Name != "" {
-						if _, ok := portForwardMap[forwardPort.Name]; ok {
-							causes = append(causes, metav1.StatusCause{
-								Type:    metav1.CauseTypeFieldValueDuplicate,
-								Message: fmt.Sprintf("Duplicate name of the port: %s", forwardPort.Name),
-								Field:   field.Child("domain", "devices", "interfaces").Index(idx).Child("ports").Index(portIdx).Child("name").String(),
-							})
-						}
-
-						if msgs := validation.IsValidPortName(forwardPort.Name); len(msgs) != 0 {
-							causes = append(causes, metav1.StatusCause{
-								Type:    metav1.CauseTypeFieldValueInvalid,
-								Message: fmt.Sprintf("Invalid name of the port: %s", forwardPort.Name),
-								Field:   field.Child("domain", "devices", "interfaces").Index(idx).Child("ports").Index(portIdx).Child("name").String(),
-							})
-						}
-
-						portForwardMap[forwardPort.Name] = struct{}{}
-					}
+					portForwardMap[forwardPort.Name] = struct{}{}
 				}
 			}
+		}
 
-			// verify that selected model is supported
-			if iface.Model != "" {
-				isModelSupported := func(model string) bool {
-					for _, m := range validInterfaceModels {
-						if m == model {
-							return true
-						}
+		// verify that selected model is supported
+		if iface.Model != "" {
+			isModelSupported := func(model string) bool {
```

Comment:
can we move the creation of the function outside the loops?

Reply:
replaced this and another occurrence with a map.